File: lab-1/set.rb
require './card.rb'
require 'colorize'

def create_deck
  deck = []

  colors = ["red", "green", "purple"]
  numbers = [1, 2, 3]
  shapes = ["diamond", "squiggle", "oval"]
  shadings = ["solid", "striped", "open"]

  colors.each do |color|
    numbers.each do |number|
      shapes.each do |shape|
        shadings.each do |shading|
          deck << Card.new(color, number, shape, shading)
        end
      end
    end
  end

  deck
end

# Finds a set in the array of cards passed in. Borrowed from https://github.com/teixeir3/Ruby-SetGame/blob/master/lib/board.rb
def find_set(cards)
  cards.combination(3).to_a.find {|x,y,z| Card.valid_set?(x,y,z)} || false
end

# Print the given array of cards in an easily-read table, one attribute per column
def display_cards(cards)
  puts
  cards.each_with_index do |card, index|
    card_color = ""
    card_color = :light_red if card.color == "red"
    card_color = :light_green if card.color == "green"
    card_color = :light_magenta if card.color == "purple"
    print "#{index}\t", "#{card.color}".colorize(card_color), "\t#{card.number}\t#{card.shape}    \t#{card.shading}\n"

  end
end


# Main game code
puts "Game Started"
deck = create_deck
playerA_score = 0
playerB_score = 0

# Randomly pick 12 cards from the deck
totalCard = 0
dealt_cards = []

  # Deal 12 cards to the table
  while totalCard < 12 
    item = deck.sample
    dealt_cards << deck.delete(item)
    totalCard = totalCard + 1
  end

  # Display the 12 drawn cards
  display_cards dealt_cards

  # Prompt the user to choose three cards then determine if they are a set
  MAXCARDS = 21
  gameOver = false
  puts "\nEnter the indices of three cards that make a set."
  puts "Player A, type your anwer as \"A #,#,#\""
  puts "Player B, type your anwer as \"B #,#,#\""
  puts "If you see no valid sets, enter 'n' to recieve three more cards."
  puts "If you need help, enter 'h' to view a hint."
  until gameOver == true
    input = gets.chomp
    if input == "n"
      #if there are no more cards in the deck, the game is over
      if deck.length == 0
        puts "Sorry there are no more cards. Game Over!"
        gameOver = true
      elsif dealt_cards.length == MAXCARDS
        puts "\nSorry, you can only have 21 cards out at a time!"
        puts "Would you like to keep looking for a set? (Enter 'y' to keep looking or 'n' to end the game)."
        keepPlaying = ""
        until keepPlaying == "y" || keepPlaying == "n"
          keepPlaying = gets.chomp
          if keepPlaying == "n"
            puts "Game Over!"
            gameOver = true
          elsif keepPlaying != "y"
            puts "Please enter either 'y' or 'n' to keep playing or quit."
          end
        end
      else
        # Print out three new random cards from the deck
        for i in 0..2
          item = deck.sample
          dealt_cards << deck.delete(item)
          totalCard = totalCard + 1
        end
        display_cards dealt_cards
      end

    elsif input == "h"
      set = find_set(dealt_cards)
      if set
        puts "There is a card that forms a set with cards #{dealt_cards.find_index(set[0])} and #{dealt_cards.find_index(set[1])}. Good luck!"
      else
        puts "There are no valid sets. Press 'n' to draw more cards! If you still can't find a set, the game is over!"
      end
    else
      #if string was not 'n' or 'h', make a string only containing numbers the user input, convert to array
      playerID = ""
      # Remove the player ID from the input
      until playerID == "A" || playerID == "B"
        playerID = (input.slice!(0)).upcase
        if playerID != "A" && playerID != "B"
          puts "ERROR: Make sure to enter either 'A' or 'B' before your set to indicate your player ID!"
          input = gets.chomp
        end
      end
      nums = input.split(',')
      nums.map!(&:to_i)
      #output error message if the user input less or more than 3 card numbers
      if nums.length != 3
        puts "ERROR: Make sure to enter exactly three card numbers with commas separating them!"
      else
        validSet = Card.valid_set?(dealt_cards[nums[0]], dealt_cards[nums[1]], dealt_cards[nums[2]])
        puts "\nCorrect. Good Job, Player #{playerID}!" if  validSet
        puts "\nIncorrect. Not a valid set. Try again!" if !validSet
        puts
        if validSet
                # Replace the cards that were selected
          if dealt_cards.length <= 12
            for i in 0..2
              item = deck.sample
              dealt_cards[nums[i]] = deck.delete(item)
            end
          else  # Remove the cards that were selected
            for i in 0..2
              card = dealt_cards[nums[i]]
              dealt_cards.delete(card)  
            end
          end

          # Add 1 to the score of the player who found the set
          playerA_score += 1 if playerID == "A"
          playerB_score += 1 if playerID == "B"
          puts "Invalid player ID entered; score not counted" if playerID != "A" && playerID != "B"

          #reprint out all the cards from the previous round
          display_cards dealt_cards

          # Display current player scores
          puts "\nPlayer A score: #{playerA_score}"
          puts "Player B score: #{playerB_score}"
          puts "There are #{deck.length} cards in the deck!"
        end
      end
    end
  end

File: lab-2/set.js
const Card = require('./card.js');
const fs = require('fs');
const { table } = require('console');
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function createDeck() {
  deck = new Array;

  var colors = ['red', 'green', 'purple'];
  var numbers = [1, 2, 3];
  var shapes = ['diamond', 'squiggle', 'oval']
  var shadings = ['solid', 'striped', 'open']

  colors.forEach(function (color) {
    numbers.forEach(function (number) {
      shapes.forEach(function (shape) {
        shadings.forEach(function (shading) {
          deck.push(new Card(color, number, shape, shading));
        });
      });
    });
  });

  return deck;
}

// Taken from https://stackoverflow.com/questions/2450954/how-to-randomize-shuffle-a-javascript-array
function shuffle(array) {
  let currentIndex = array.length, randomIndex;

  // While there remain elements to shuffle...
  while (currentIndex != 0) {

    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex--;

    // And swap it with the current element.
    [array[currentIndex], array[randomIndex]] = [
      array[randomIndex], array[currentIndex]];
  }

  return array;
}

function displayCards(cards) {
  for (var index = 0; index < cards.length; index++) {
    console.log(`${index}\t${cards[index].color}\t${cards[index].number}\t${cards[index].shape}    \t${cards[index].shading}`);
  }
}

/*=========================
        Main Code 
===========================*/
// Create and shuffle the full card deck
var deck = createDeck();
var totalTableCards = 0;
var playerA_score = 0;
var playerB_score = 0;
shuffle(deck);

// Deal the first 12 cards of the shuffled array to the table
var tableCards = new Array(12);
for (var i = 0; i < 12; i++) {
  tableCards[i] = deck.shift();
  totalTableCards = totalTableCards + 1;
}
displayCards(tableCards);

var gameOver = false;
const maxCards = 21;

console.log('\nEnter the indices of three cards that make a set.');
console.log('Player A, type your choice as "A #,#,#');
console.log('Player B, type your choice as "B #,#,#');
console.log('If you se no valid sets, enter "n" to receive three more cards.');
console.log('If you need help, enter "h" to view a hint.');

while (gameOver == false) {
  if(/*user click new card*/){
    if(deck.length == 0) {
      window.alert('Sorry there are no more cards. Game Over!');
    } else if(tableCards.length == maxCards) {
      window.alert('\nSorry, you can only have 21 cards out at a time!');
      window.alert('Would you like to keep looking for a set?');
      if(/*the user click yes*/){
        console.log("The game is continued");
      } else if(/* the user click no*/){
        gameOver == true;
        console.log("\nGame Over! The final score is:");
        console.log("Player A score:" + playerA_score);
        console.log("Player B score:" + playerB_score);
        if( playerA_score > playerB_score){
          console.log("Player A wins!");
        } else if(playerB_score > playerA_score){
          console.log("Player B wins!");
        } else{
          console.log("The game ends in a draw!");
        }
      }
    } else{
        for(var i = 0; i <= 2; i++){
          totalTableCards = totalTableCards + i;
          tableCards[totalTableCards] = deck.shift();
      }
      displayCards(tableCards);
    }
  } else if(/*user clicks hint*/) {

  }
}
